plugins/unreal/XJGame/Plugins/XJMusicPlugin/Source/XJMusicPlugin/Private/Test/PrototypeActor.cpp

```diff
@@ -0,0 +1,286 @@
+{
+	if (!XjEngine)
+	{
+		return;
+	}
+
+	GEngine->AddOnScreenDebugMessage(-1, DeltaTime, FColor::Green, FString("Chain micros: ") + AtChainMicros.ToString(), false);
+
+	std::set<ActiveAudio> ReceivedAudios = XjEngine->runCycle(AtChainMicros.GetMicros());
+
+	for (const ActiveAudio& Audio : ReceivedAudios)
+	{
+		FString WaveKey = Audio.getAudio()->waveformKey.c_str();
+		FString Name = Audio.getAudio()->name.c_str();
+
+		long TransientMicros = Audio.getAudio()->transientSeconds * MICROS_PER_SECOND;
+		long LengthMicros = Audio.getPick()->lengthMicros;
+
+		TimeRecord StartTime = Audio.getStartAtChainMicros();
+		TimeRecord EndTime = Audio.getStopAtChainMicros().value();
+
+		FAudioPlayer AudioPlayer;
+		AudioPlayer.StartTime = StartTime;
+		AudioPlayer.EndTime = EndTime;
+		AudioPlayer.Name = Name;
+		AudioPlayer.Id = WaveKey;
+
+		if (AudiosLookup.Contains(StartTime))
+		{
+			TArray<FAudioPlayer>& SavedAudios = AudiosLookup[StartTime];
+
+			bool SkipAdding = false;
+
+			for (FAudioPlayer& SavedAudio : SavedAudios)
+			{
+				if (SavedAudio.Id != WaveKey)
+				{
+					continue;
+				}
+
+				if (SavedAudio.StartTime == AudioPlayer.StartTime)
+				{
+					SkipAdding = true;
+
+					if (!(SavedAudio.EndTime == AudioPlayer.EndTime))
+					{
+						//Update end time then
+					}
+				}
+			}
+
+			if (!SkipAdding)
+			{
+				SavedAudios.Add(AudioPlayer);
+			}
+		}
+		else
+		{
+			AudiosLookup.Add(StartTime, { AudioPlayer });
+			AudiosKeys.Add(StartTime);
+		}
+
+		if (DebugViewAudioToTime.Contains(Name))
+		{
+			DebugViewAudioToTime[Name].Add(AudioPlayer);
+		}
+		else
+		{
+			DebugViewAudioToTime.Add(Name, { AudioPlayer });
+		}
+
+		if (DebugViewTimeToAudio.Contains(StartTime))
+		{
+			DebugViewTimeToAudio[StartTime].Add(Name);
+		}
+		else
+		{
+			DebugViewTimeToAudio.Add(StartTime, { Name });
+		}
+	}
+
+	AudiosKeys.Sort();
+}
+
+void APrototypeActor::Tick(float DeltaTime)
+{
+	Super::Tick(DeltaTime);
+
+	if (!HasSegmentsDubbedPastMinimumOffset() && IsWithinTimeLimit())
+	{
+		RunXjOneCycleTick(DeltaTime);
+	}
+
+	if (!XjMusicInstanceSubsystem)
+	{
+		return;
+	}
+
+	FString PlayingAudios = "Playing:\n";
+
+	for (TimeRecord Time : AudiosKeys)
+	{
+		if (!AudiosLookup.Contains(Time))
+		{
+			continue;
+		}
+
+		TArray<FAudioPlayer>& Audios = AudiosLookup[Time];
+		
+		Audios.RemoveAll([this, &PlayingAudios](FAudioPlayer& Element)
+			{
+				if (Element.StartTime > AtChainMicros.GetMicros())
+				{
+					return false;
+				}
+
+				if (Element.EndTime <= AtChainMicros.GetMicros())
+				{
+					XjMusicInstanceSubsystem->StopAudioByName(Element.Id);
+					return true;
+				}
+
+				PlayingAudios += FString::Printf(TEXT("%s start - %f  end - %f\n"), *Element.Name, Element.StartTime.GetSeconds(), Element.EndTime.GetSeconds());
+
+				if (Element.bIsPlaying)
+				{
+					return false;
+				}
+
+				Element.bIsPlaying = true;
+
+				XjMusicInstanceSubsystem->PlayAudioByName(Element.Id);
```

Comment:
I'm wondering why this is called here with no second parameter value for **OverrideStartTime**

Reply:
Nvm, figured it out. See my comments on the PlayAudioByName function.